components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/DefaultIDTokenBuilder.java

```diff
@@ -792,5 +805,43 @@ private boolean isValidIdToken(JWTClaimsSet jwtClaimsSet) {
     private long getIDTokenExpiryInMillis() {
         return OAuthServerConfiguration.getInstance().getOpenIDConnectIDTokenExpiryTimeInSeconds() * 1000L;
     }
-}
 
+    private void setAdditionalClaims(OAuthTokenReqMessageContext tokenReqMsgCtxt,
+                                     OAuth2AccessTokenRespDTO tokenRespDTO, JWTClaimsSet jwtClaimsSet)
+            throws IdentityOAuth2Exception {
+        List<ClaimProvider> claimProviders = getClaimProviders();
+        if (CollectionUtils.isNotEmpty(claimProviders)) {
+            for (ClaimProvider claimProvider : claimProviders) {
+                Map<String, Object> additionalJWTClaims =
+                        claimProvider.getAdditionalClaims(tokenReqMsgCtxt, tokenRespDTO);
+                setAdditionalClaimSet(jwtClaimsSet, additionalJWTClaims);
+            }
+        }
+    }
+
+    private void setAdditionalClaims(OAuthAuthzReqMessageContext authzReqMessageContext,
+                                     OAuth2AuthorizeRespDTO authorizeRespDTO, JWTClaimsSet jwtClaimsSet)
+            throws IdentityOAuth2Exception {
+        List<ClaimProvider> claimProviders = getClaimProviders();
+        if (CollectionUtils.isNotEmpty(claimProviders)) {
+            for (ClaimProvider claimProvider : claimProviders) {
+                Map<String, Object> additionalJWTClaims =
+                        claimProvider.getAdditionalClaims(authzReqMessageContext, authorizeRespDTO);
+                setAdditionalClaimSet(jwtClaimsSet, additionalJWTClaims);
+            }
+        }
+    }
+
+    private List<ClaimProvider> getClaimProviders() {
+        return OpenIDConnectServiceComponentHolder.getClaimProviders();
+    }
+
+    private void setAdditionalClaimSet(JWTClaimsSet jwtClaimsSet, Map<String, Object> additionalJWTClaimsSet) {
+        jwtClaimsSet.setAllClaims(additionalJWTClaimsSet);
+        if (log.isDebugEnabled()) {
+            for (Map.Entry<String, Object> entry : additionalJWTClaimsSet.entrySet()) {
+                log.debug("Additional claim found :" + entry.getKey() + ": " + entry.getValue());
```

Comment:
"Additional claim found :" + entry.getKey() + ": " + entry.getValue() -->
"Additional claim added to JWTClaimSet, key: " + entry.getKey() + ", value: " + entry.getValue()